### PR TITLE
BPH effect update (data split and fade-in)

### PR DESCRIPTION
**Pull request by mikewin-climsci**
# Description

## Purpose of this PR

- This PR adds two new features to the BPH effect of afforestation in the (dynamic_may20) realization of the 32_forestry module. Default is OFF.
- 1) The input dataset is now split up into the BPH effect and the climate data (TCRE) needed to translate it to it's carbon equivalent to allow for more flexibility
- 2) The boolean ageclass switch is replaced by a more realistic fade-in between ac10 and ac30

## Performance loss/gain from current default behavior

[PR_BPH_split.pdf](https://github.com/magpiemodel/magpie/files/4930967/PR_BPH_split.pdf)
PDF shows the afforestation area affected by the PR. Depicted are the develop of the 16.07. with and without the PR (default BPH off), and experiment run with the BPH effect set to on.

- %performance loss = ~+5% runtime 100*PR/default

## Type of change

- [ ] Bug fix (Change which fixes an issue).
- [X] New feature (Change which adds functionality).
- [ ] Minor change (Change which does not modify core model code i.e., in /modules).
- [ ] Major change (fix or feature that would change current model behavior).

# How Has This Been Tested?

- [x] Runs using starting script which tests current defaults (using `test_runs.R`).
- [X] Runs using starting script which successfully runs the same scenarios as in `test_runs.R` but with the changes from PR.

# Default configuration additions/changes:

- [ ] Realizations:
- [X] Scalars/Constants:
- [X] Other changes: Additional preloop equations and input files

# Checklist:

- [X] Self-review of my own code.
- [X] Added changes to `CHANGELOG.md`
- [X] No hard coded numbers and cluster/country/region names.
- [X] The code doesn't contain declared but unused parameters or variables.
- [X] In-code comments added including documentation comments.
- [-] Compiled and checked resulting documentation with `goxygen::goxygen()` for the new/updated parts.
- [-] Changes to `magpie4` R library for post processing of model output (ideally backward compatible).

## Special comments/warnings

- Default is BPH OFF.

### CODE REVIEW

File: modules/32_forestry/dynamic_may20/input.gms [hunk 1]
```diff
@@ -2,13 +2,13 @@ $setglobal c32_timber_plantations  plantations
 * option: natveg, plantations
 $setglobal c32_aff_mask  noboreal
 * options: unrestricted, noboreal, onlytropical
-$setglobal c32_aff_policy  npi
```
Comment (flohump) [inline modules/32_forestry/dynamic_may20/input.gms:5]: Please revert back to the default setting "npi"

File: modules/32_forestry/dynamic_may20/preloop.gms [hunk 2]
```diff
@@ -158,9 +158,27 @@ elseif s32_initial_distribution = 1,
...
+*fade-in from ac10 to ac30. First effect in ac10 (ord 3), last effect in ac30 (ord 7).
+ac_bph(ac) = no;
+ac_bph(ac) = yes$(ord(ac) >= 3 AND ord(ac) <= 7);
+display ac_bph;
+
 p32_aff_bgp(j,ac) = 0;
-p32_aff_bgp(j,"%c32_bgp_ac%") = f32_aff_bgp(j,"%c32_aff_bgp%");
+p32_tcre_glo(j) = 0;
+if(s32_tcre_local = 1,
+	p32_aff_bgp(j,ac_bph) = f32_aff_bgp(j,"%c32_aff_bgp%")/f32_tcre(j,"%c32_tcre_ctrl%")/card(ac_bph);
```
Comment (flohump) [inline modules/32_forestry/dynamic_may20/preloop.gms:173]: Are you sure that you don't need brackets around f32_aff_bgp(j,"%c32_aff_bgp%")/f32_tcre(j,"%c32_tcre_ctrl%")?
Just double check if your calculation is correct

File: modules/32_forestry/dynamic_may20/preloop.gms [hunk 3]
```diff
@@ -158,9 +158,27 @@ elseif s32_initial_distribution = 1,
...
+else
+*m_weightedmean returns a global value, which is then used assigned to all j. We use land area as weight.
+  p32_tcre_glo(j2) = m_weightedmean(f32_tcre(j,"%c32_tcre_ctrl%"),sum(land, pcm_land(j,land)),j)
+  p32_aff_bgp(j,ac_bph) = f32_aff_bgp(j,"%c32_aff_bgp%")/p32_tcre_glo(j)/card(ac_bph)
```
Comment (flohump) [inline modules/32_forestry/dynamic_may20/preloop.gms:177]: Same here. Calculation correct?
Did you check the resutling values in p32_aff_bgp. Do they make sense?

File: modules/32_forestry/dynamic_may20/input.gms (hunk 1 above)
Comment (mikewin-climsci) [reply modules/32_forestry/dynamic_may20/input.gms]: Changed default aff_policy back to npi (commit of 28.07.2020)

File: modules/32_forestry/dynamic_may20/preloop.gms (hunk 3 above)
Comment (mikewin-climsci) [reply modules/32_forestry/dynamic_may20/preloop.gms:177]: I checked the calculation and it works as intended.

- A back-of-the-envelope calculation approximates the BGP effect at around 0-3 tc/ha (already divided into ageclasses).

- I checked an example area (LAM_80): 
Intended results: Local = 2.45 ; Global = 2.21
Results in .lst: Local 2.45 ; Global = 2.21

Output from runs in /p/projects/landuse/users/windisch/develop_bph/magpie/output:
(BPH_calctest_windisch_SSP2-POL_PBL_2019_withPR_local;BPH_calctest_windisch_SSP2-POL_PBL_2019_withPR_global)

File: modules/32_forestry/dynamic_may20/input.gms [hunk 4]
```diff
@@ -6,9 +6,9 @@ $setglobal c32_aff_policy  npi
 * options: none, npi, ndc
 
 $setglobal c32_aff_bgp  nobgp
-* options: ann,djf,jja,nobgp
-$setglobal c32_bgp_ac  ac10
-* options: ac0,ac10,ac20,ac30
+* options: ann,nobgp
+$setglobal c32_tcre_ctrl  ann_TCREmean
```
Comment (tscheypidi) [inline modules/32_forestry/dynamic_may20/input.gms:10]: c32_tcre_ctrl needs to show up in default.cfg as well!

Comment (mikewin-climsci) [reply modules/32_forestry/dynamic_may20/input.gms:10]: Added the missing switch with description to the default.cfg:

```
Switch for the different TCRE estimates which are used to
translate the BGP temperature effect into their carbon equivalent.
The estimates encompass the ensemble mean +/- SD of 20 CMIP5 models.
Ceq ~ BGP/TCRE - Hence, high TCRE decreases the BGP effect.
("ann_TCREmean"): Ensemble mean
("ann_TCREhigh"): Ensemble mean + SD
("ann_TCRElow"): Ensemble mean - SD
cfg$gms$c32_tcre_ctrl <- "ann_TCREmean"
```
